### PR TITLE
🐛 Stub out elements that fail to upgrade

### PR DESCRIPTION
**Pull request by jridgewell**
I think this is what's causing #25381. If a component, eg `<amp-link-rewritter>`, accesses `this.getAmpDoc()` to get a service in it's constructor, it'll error out.

If that happens, then `this.implementation_` is still `undefined`. That causes a cascade of bugs. So, if the component fails to construct, we're now going to turn it into a stubbed `BaseElement` and pretend it doesn't exist on the page.

### CODE REVIEW

File: src/custom-element.js [hunk 1]
```diff
@@ -328,8 +333,19 @@ function createBaseCustomElementClass(win) {
...
+      }
+      if (!this.isConnected_) {
+        throw dev().createError(
+          this.tagName.toLowerCase(),
+          'attempted to get ampdoc before being connected'
+        );
+      }
+      const win = toWin(this.ownerDocument.defaultView);
+      const ampdocService = Services.ampdocServiceFor(win);
+      this.ampdoc_ = ampdocService.getAmpDoc(this);
```
Comment (dvoytenko) [inline src/custom-element.js]: I think we just need `Services.ampdoc(this)` now. Just an API refresh, so we can simplify this a bit.

File: src/service.js [hunk 2]
```diff
@@ -346,6 +346,7 @@ export function getParentWindowFrameElement(node, opt_topWin) {
  */
 export function getAmpdoc(nodeOrDoc) {
   if (nodeOrDoc.nodeType) {
+    // TODO: isConnected check on document?
```
Comment (dvoytenko) [inline src/service.js]: I believe `isConnected` is checked in the `ampdoc-impl`.

File: src/custom-element.js (hunk 1 above)
Comment (jridgewell) [reply src/custom-element.js]: Done.

File: src/service.js (hunk 2 above)
Comment (jridgewell) [reply src/service.js]: This is for the case where you pass a `Node` to a service getter, not an `ampdoc`. Doing `node.ownerDocument` is not safe in cases where we inject a node into an FIE, so we need to wait till it's connected first.

File: src/custom-element.js [hunk 3]
```diff
@@ -301,6 +286,26 @@ function createBaseCustomElementClass(win) {
...
+      let Ctor =
+        win.__AMP_EXTENDED_ELEMENTS &&
+        win.__AMP_EXTENDED_ELEMENTS[this.elementName()];
+      if (getMode().test && nonStructThis['implementationClassForTesting']) {
+        Ctor = nonStructThis['implementationClassForTesting'];
+      }
+      devAssert(Ctor);
+      try {
+        /** @private {!./base-element.BaseElement} */
+        this.implementation_ = new Ctor(this);
```
Comment (jridgewell) [inline src/custom-element.js]: I'm also happy to do this in the CE's `connectedCallback` after getting the ampdoc. There should be little downside.

Comment (dvoytenko) [reply src/custom-element.js]: Let's do that, but in a separate PR.

**Comment by jridgewell** [comment]
Closing for https://github.com/ampproject/amphtml/pull/25491.

**Pull request closed without merging**